common/static/js/vendor/tinymce/BUILD_README.txt

```diff
@@ -1,20 +1,55 @@
+1. Download the desired version from https://github.com/tinymce/tinymce/tags
+2. If it’s a major update, follow the official migration doc and update the codebase as needed.
+3. Setup the codemirror-plugin as per the instruction below.
+4. Find all the EDX specific changes in the currently used version of tinymce by searching for the string "EDX" in the vendor/js/tinymce dir.
+5. Merge the EDX specific changes with the newly downloaded version.
+6. Follow the instructions given below to create the new version of js/tinymce.full.min.js
 
-Instructions for updating tinymce to a newer version:
+# Instruction for setting codemirror-plugin
 
-1. Download the desired version from https://github.com/tinymce/tinymce/releases
-2. Find all the EDX specific changes that were made to the currently used version of tinymce by searching for
-   the string "EDX" in this directory.
-3. Merge the EDX specific changes with the new version.
-4. Follow the instructions above for creating the new version of js/tinymce.full.min.js
+1. Download the tinymce-codemirror-plugin from https://gitlab.com/tinymce-plugins/tinymce-codemirror
+2. Open terminal in the downloaded plugin directory and run the following commands:
+    ```
+     npm install
+     npm run prepublish (This command will generate the minified file in the plugin directory)
+    ```
+3. Remove the tinymce-codemirror/plugins/codemirror/codemirror-4.8 directory
+4. Move the tinymce-codemirror/plugins directory to `common/static/js/vendor/tinymce/js/plugins/` directory.
+5. Apply EDX specific changes in the existing code to the `plugin.js` and `source.html` files.
+6. Install [uglify-js](https://www.npmjs.com/package/uglify-js) and generate `plugin.min.js`
+    ```
+    cd common/static/js/vendor/tinymce/js/plugins/codemirror/
+    uglify plugin.js -m -o plugin.min.js
+    ```
+
+# Instructions for creating js/tinymce.full.min.js
+
+The following uses the version 5.5.1 as a reference. Change your filenames depending the version you have downloaded.
+
+1. Unzip the zip file downloaded from Github.
+    ```
+    unzip tinymce-5.5.1.zip
+    ```
+2. Open terminal and change directory to the newly downloaded tinymce.
+    ```
+    cd tinymce-5.5.1
+    ```
+3. Build TinyMCE using Yarn. this will create multiple zip files in the `dist` directory.
+    ```
+    yarn && yarn build
+    ```
+4. Unzip the dev bundle to the edx-platform's vendor directory.
+    ```
+    unzip dist/tinymce_5.5.1_dev.zip -d /path/to/edx-platform/common/static/js/vendor/
+    ```
+5. Remove the unnecessary files in `/path/to/edx-platform/common/static/js/vendor/tinymce` like `package.json`, `yarn.lock`...etc.,
+6. Generate a bundled version of the TinyMCE with all the plugins using the following command
+    ```
+    cd common/static/js/vendor/tinymce/js/tinymce
+    cat tinymce.min.js */*/*.min.js plugins/emoticons/js/emojis.min.js > tinymce.full.min.js
```

Comment:
@Agrendalath I tried the pattern `*/**/*.min.js` in the `cat` it still missed out the `emojis.min.js` file for some reason. So I have kept it as it is.

About using the `find` command, I didn't find any problems with it. I just found a simpler command [here](https://github.com/openedx/edx-platform/pull/3155/commits/9d9b2b8e185f5909abd2abb618aa4d5a7c6fce09) when going through the commits of #3155 and used it as it is much easier to read & understand and does the same task.

Reply:
@tecoholic,
> I tried the pattern */**/*.min.js in the cat it still missed out the emojis.min.js file for some reason. So I have kept it as it is.

Ah, right I forgot that the globstar is disabled by default in bash (it needs to be enabled with `shopt -s globstar` to work this way).

> About using the find command, I didn't find any problems with it. I just found a simpler command [here](https://github.com/openedx/edx-platform/pull/3155/commits/9d9b2b8e185f5909abd2abb618aa4d5a7c6fce09) when going through the commits of https://github.com/openedx/edx-platform/pull/3155 and used it as it is much easier to read & understand and does the same task.

I still believe that producing a deterministic (in this case system-agnostic) output is better. We should not assume that people will remember the manual work required when any changes are applied to the TinyMCE library or any of its plugins. If we had the same result on every machine, we could easily verify whether this file has been updated.
I'll let you choose which command you would like to keep, though.

Reply:
@Agrendalath I have been scratching my head over how this can be done. 
* The globbing pattern in a shell always returns a [consistent list](https://serverfault.com/questions/122737/in-bash-are-wildcard-expansions-guaranteed-to-be-in-order)
* But the output will [vary between systems](https://unix.stackexchange.com/questions/368318/does-the-bash-star-wildcard-always-produce-an-ascending-sorted-list) base on LC_COLLATE, LOCALE values. So consistent between machines cannot be guaranteed.
* The different shells might also implement this differently. I tested with Bash and Fist and got different outputs for `tinymce.full.min.js`

With this in mind I think storing the list of .min.js files explicitly as a file and then using something like `xargs cat < filelist > tinymce.full.min.js`might provide a consistent output irrespective of the environment. I have tested this Bash and Fish and the output are consistent. However this comes with the drawback of keeping an extra file which might not be maintained, if a plugin was added or removed.

What do you think?

Reply:
@tecoholic, I'm not sure that I follow… Why don't we simply use `| LC_ALL=C sort -r |`?

Reply:
@Agrendalath I over-complicated it as I needed to maintain the order as it would load in browser 
1. tinymce
2. plugins (the order of plugins themselves don't matter)
3. plugins' subscripts
and also make sure the glob pattern can ignore jquery.tinymce.min.js and the existing `tinymce.full.min.js` in the folder.

I think the following produces consistent output and I have update the README and the 'tinymce.full.min.js` in 9f139136c85961b2afb1a287bb26828ab19eb773
```
LC_ALL=C cat tinymce.min.js */*/*.min.js plugins/emoticons/js/emojis.min.js > tinymce.full.min.js
```